File: OSX/SafeBoardSandbox/TodoMainViewController.m
//
//  SBTodoItemsMainViewController.m
//  SafeBoardSandbox
//
//  Created by Vladimir Kukushkin on 12/23/15.
//  Copyright © 2015 Vladimir Kukushkin. All rights reserved.
//

#import "TodoMainViewController.h"

@implementation TodoMainViewController_objc

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [m_itemsArrayController addObserver:self
                             forKeyPath:@"arrangedObjects.todoTitle"
                                options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld)
                                context:NULL];
    [m_itemsArrayController addObserver:self
                             forKeyPath:@"arrangedObjects.todoDescription"
                                options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld)
                                context:NULL];
    [m_itemsArrayController addObserver:self
                             forKeyPath:@"arrangedObjects.todoDueDate"
                                options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld)
                                context:NULL];
    
    self.manager = [TodoManager new];
    [self.manager connect:self];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(NSArrayController*)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    
    if(self.manager.items.count <= object.selectionIndex)
        return;
    
    TodoItem * selectedItem = [m_itemsArrayController.arrangedObjects objectAtIndex:object.selectionIndex];
    
    if([keyPath isEqualToString:@"arrangedObjects.todoTitle"]) {
        [self.manager updateTitleforObjectAtIndex:object.selectionIndex withValue:selectedItem.todoTitle];
    }
    if([keyPath isEqualToString:@"arrangedObjects.todoDescription"]) {
        [self.manager updateDescriptionforObjectAtIndex:object.selectionIndex withValue:selectedItem.todoDescription];
    }
    if([keyPath isEqualToString:@"arrangedObjects.todoDueDate"]) {
        [self.manager updateDueDateforObjectAtIndex:object.selectionIndex withValue:selectedItem.todoDueDate];
    }
}

- (IBAction)add:(id)sender {
    [self.manager createObject];
}

- (IBAction)remove:(id)sender {
    [self.manager removeObjectAtIndex:m_itemsArrayController.selectionIndex];
}

@end

File: OSX/SafeBoardSandbox/TodoMainViewController.swift
//
//  TodoMainViewController.swift
//  SafeBoardSandbox
//
//  Created by Vladimir Kukushkin on 12/24/15.
//  Copyright © 2015 Kaspersky Lab. All rights reserved.
//

//
// Notice:
// To use this controller and _swift project, you should change class of Main View Controller object in MainMenu.xib
//

import Foundation
import Cocoa
import TodoManagerModel

class TodoMainViewController_swift : NSViewController {
    
    @IBOutlet var m_itemsArrayController : NSArrayController!
    
    @objc dynamic var manager: TodoManager {
        return TodoManager.shared()
    }
    
    override func viewDidLoad() {
        
        let options = NSKeyValueObservingOptions([.new, .old])
        
        m_itemsArrayController.addObserver(self, forKeyPath:"arrangedObjects.todoTitle", options:options, context:nil)
        m_itemsArrayController.addObserver(self, forKeyPath:"arrangedObjects.todoDescription", options:options, context:nil)
        m_itemsArrayController.addObserver(self, forKeyPath:"arrangedObjects.todoDueDate", options:options, context:nil)
        
        self.manager.connect(self);
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if let arrayController = object as? NSArrayController {
            if(self.manager.items.count <= (arrayController as AnyObject).selectionIndex) {
                return;
            }
            
            if let selectedItem = (m_itemsArrayController.arrangedObjects as! NSArray).object(at:arrayController.selectionIndex) as? TodoItem {
                if(keyPath == "arrangedObjects.todoTitle") {
                    self.manager.updateTitleforObject(at: arrayController.selectionIndex, withValue:selectedItem.todoTitle)
                }
                if(keyPath == "arrangedObjects.todoDescription") {
                    self.manager.updateDescriptionforObject(at: arrayController.selectionIndex, withValue:selectedItem.todoDescription);
                }
                if(keyPath == "arrangedObjects.todoDueDate") {
                    self.manager.updateDueDateforObject(at: arrayController.selectionIndex, withValue:selectedItem.todoDueDate);
                }
            }
        }
    }
    
    @IBAction func add(_ sender: AnyObject?) {
        self.manager.createObject()
    }
    
    @IBAction func remove(_ sender: AnyObject?) {
        self.manager.removeObject(at: m_itemsArrayController.selectionIndex)
    }
}
